express-notes-api/index.js

```diff
@@ -0,0 +1,85 @@
+const express = require('express');
+const app = express();
+const middleware = express.json();
+const data = require('./data.json');
+const fs = require('fs');
+
+app.listen(3000, () => {
+  // eslint-disable-next-line no-console
+  console.log('Express server listening on port 3000');
+});
+
+app.use(middleware);
+
+app.get('/api/notes', (req, res) => {
+  const arr = [];
+  for (const keys in data.notes) {
+    arr.push(data.notes[keys]);
+  }
+  res.status(200).json(arr);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  if (req.params.id < 0) {
+    res.status(400).json({ error: 'id must be a positive integer' });
+  } else if (!data.notes[req.params.id]) {
+    res.status(404).json({ error: `cannot find note with id ${req.params.id}` });
+  } else {
+    res.status(200).json(data.notes[req.params.id]);
+  }
+});
+
+app.post('/api/notes', (req, res) => {
+  if (!req.body) {
+    res.status(400).json({ error: 'content is a required field' });
+  } else {
+    Object.assign(data.notes, { [data.nextId]: req.body });
+    data.notes[data.nextId].id = data.nextId;
+    data.nextId++;
+    const stringifyData = JSON.stringify(data, null, 2);
+    fs.writeFile('data.json', stringifyData, err => {
+      if (err) {
+        res.status(500).json({ error: 'An unexpected error occurred.' });
+      }
+    });
+    res.status(201).send(req.body);
```

Comment:
Since `fs.writeFile` is asynchronous, you can't send a success response until you know that it has completed. Use an `else` block in your `writeFile` callback. In case of error, send the `500`, otherwise, you can send the `201`